### PR TITLE
Cps/transonic calculate rhs

### PR DESCRIPTION
**Pull request by eloisabaez**
Functions and test to calculate the left hand side and right hand side for a supersonic element which follow the below equations:

RHS:
Screen Shot 2020-06-25 at 3 23 05 PM
LHS:
image

Based the residual definition of A.2.7 in Fully Simultaneous Coupling of the Full Potential Equation and the Integral Boundary Layer Equations in Three Dimensions by Brian Allen Nishida (MIT, 1996)

### CODE REVIEW

File: applications/CompressiblePotentialFlowApplication/custom_elements/transonic_perturbation_potential_flow_element.cpp [hunk 1]
```diff
@@ -953,6 +1032,44 @@ void TransonicPerturbationPotentialFlowElement<TDim, TNumNodes>::AssignRightHand
...
+    const double density = PotentialFlowUtilities::ComputeUpwindedDensity<TDim, TNumNodes>(velocity, upwindVelocity, rCurrentProcessInfo);
+        
+    const BoundedVector<double, TNumNodes> DNV = prod(data.DN_DX, velocity);
+    BoundedVector<double, TNumNodes + 1> DNV_extended;
+    DNV_extended.clear();
+    for (int i = 0; i < TNumNodes; i++) {
+        DNV_extended[i] = DNV[i];
+    }
+
+    BoundedMatrix<double, TNumNodes, TNumNodes> common_part_lhs = data.vol * density * prod(data.DN_DX, trans(data.DN_DX));
```
Comment (inigolcanalejo) [inline applications/CompressiblePotentialFlowApplication/custom_elements/transonic_perturbation_potential_flow_element.cpp]:
```suggestion
    BoundedMatrix<double, TNumNodes, TNumNodes> linear_term = data.vol * density * prod(data.DN_DX, trans(data.DN_DX));
```

File: applications/CompressiblePotentialFlowApplication/custom_elements/transonic_perturbation_potential_flow_element.cpp [hunk 2]
```diff
@@ -614,6 +622,77 @@ void TransonicPerturbationPotentialFlowElement<TDim, TNumNodes>::CalculateRightH
...
+
+    upwind_factor_options[1] = PotentialFlowUtilities::ComputeUpwindFactor<TDim, TNumNodes>(local_mach_number_squared, rCurrentProcessInfo);
+    upwind_factor_options[2] = PotentialFlowUtilities::ComputeUpwindFactor<TDim, TNumNodes>(upwind_mach_number_squared, rCurrentProcessInfo);
+
+    const size_t upwind_factor_case = PotentialFlowUtilities::ComputeUpwindFactorCase<TDim, TNumNodes>(upwind_factor_options);
+    
+    double DrhoDu2 = 0.0;
+    double DrhoDu2_up = 0.0;
+
+    if (upwind_factor_case == 0) { // subsonic, not inlet
```
Comment (inigolcanalejo) [inline applications/CompressiblePotentialFlowApplication/custom_elements/transonic_perturbation_potential_flow_element.cpp]:
```suggestion
    if (upwind_factor < 0.0) { // subsonic, not inlet
```

File: applications/CompressiblePotentialFlowApplication/custom_elements/transonic_perturbation_potential_flow_element.cpp [hunk 3]
```diff
@@ -614,6 +622,77 @@ void TransonicPerturbationPotentialFlowElement<TDim, TNumNodes>::CalculateRightH
...
+        // gets [TNumNodes + 1, TNumNodes + 1] size matrix
+        CalculateLeftHandSideSubsonicElement(rLeftHandSideMatrix, rCurrentProcessInfo);
+        return;
+    }
+    else if (upwind_factor_case == 1) { // supersonic, accelerating
```
Comment (inigolcanalejo) [inline applications/CompressiblePotentialFlowApplication/custom_elements/transonic_perturbation_potential_flow_element.cpp]:
```suggestion
    else if (upwind_factor >= upstream_upwind_factor) { // supersonic, accelerating
```

File: applications/CompressiblePotentialFlowApplication/custom_elements/transonic_perturbation_potential_flow_element.cpp [hunk 4]
```diff
@@ -614,6 +622,77 @@ void TransonicPerturbationPotentialFlowElement<TDim, TNumNodes>::CalculateRightH
...
+        // density derivatives
+        DrhoDu2 = PotentialFlowUtilities::ComputeUpwindedDensityDerivativeWRTVelocitySquaredSupersonicAccelerating<TDim, TNumNodes>(
+            velocity, local_mach_number_squared, upwind_mach_number_squared, rCurrentProcessInfo);
+        DrhoDu2_up = PotentialFlowUtilities::ComputeUpwindedDensityDerivativeWRTUpwindVelocitySquaredSupersonicAccelerating<TDim, TNumNodes>(
+            local_mach_number_squared, upwind_mach_number_squared, rCurrentProcessInfo);
+    }
+    else if (upwind_factor_case == 2) { // supersonic, deaccelerating
```
Comment (inigolcanalejo) [inline applications/CompressiblePotentialFlowApplication/custom_elements/transonic_perturbation_potential_flow_element.cpp]:
```suggestion
    else { // supersonic, deaccelerating
```

**Review by inigolcanalejo — CHANGES_REQUESTED** [review]
Only the minor comments we talked about. I see that you have added tests for the lhs and rhs for the supersonic accelerating case :+1: could you also add ones for the supersonic decelerating case?

**Comment by eloisabaez** [comment]
Hi Iñigo, I think I have addressed all the comments now. The only thing which I did a bit differently is I just checked the Mach numbers instead of the upwind factors in the CalculateLeftHandSideNormalElement function.

**Review by inigolcanalejo — APPROVED** [review]
Looks good :+1: Thanks for addressing the suggestions. I would say the code is ready to start with the validation. 

I think it is also a good idea to first update the python test you have in order to trigger your transonic implementation in some of the elements? You could just increase the free stream mach number. You can do this in another PR.

**Comment by inigolcanalejo** [comment]
> Looks good +1 Thanks for addressing the suggestions. I would say the code is ready to start with the validation.
> 
> I think it is also a good idea to first update the python test you have in order to trigger your transonic implementation in some of the elements? You could just increase the free stream mach number. You can do this in another PR.

The reason for making the test is that most probably you will have to change your code during the validation and it is always good to have a test with an example that you can quickly run and check whether it breaks or not

**Pull request merged**